File: training/feedback/feedback_engine.py
#!/usr/bin/env python3
"""KUMPAS Phase 6 — corrective feedback engine (Python reference).

Compares a learner's landmark sequence against the class gold standard and
produces structured feedback on four dimensions (PRD §1): TIMING, MOTION,
HANDSHAPE, ORIENTATION. This is the pedagogical core — the Kotlin port in
app/lib/feedback_engine must stay behaviourally identical to this reference.

Input sequences are (30, 258) no_face frames, already normalized by the
preprocessing pipeline (mid-hip centered, torso scaled) — so position
comparisons are signer-size- and camera-distance-invariant.

Pipeline:
  1. DTW-align learner to gold on wrist trajectories (timing-invariant view).
  2. TIMING     — warping-path slope: sustained tempo deviation.
  3. MOTION     — mean aligned wrist-position error + dominant displacement
                  direction of the error.
  4. HANDSHAPE  — per-aligned-frame finger configuration relative to the
                  wrist (translation-invariant), per-finger error.
  5. ORIENTATION— palm-plane normal angle between learner and gold.

Each dimension yields score in [0, 1] (0 = matches gold). Scores above
per-dimension thresholds become FeedbackItem objects with human-readable
prompts, worst first. Thresholds are provisional pending expert calibration.
"""

import json
from dataclasses import dataclass, field, asdict

import numpy as np

# ---- feature layout (no_face, 258) ----
POSE_N = 33
L_WRIST_P, R_WRIST_P = 15, 16          # pose wrist landmark indices
LHAND0, RHAND0 = 132, 132 + 63          # hand blocks, 21*3 each
WRIST_H, INDEX_MCP, PINKY_MCP = 0, 5, 17
FINGERS = {  # hand landmark indices per finger (MCP..TIP)
    "thumb": [1, 2, 3, 4], "index": [5, 6, 7, 8], "middle": [9, 10, 11, 12],
    "ring": [13, 14, 15, 16], "pinky": [17, 18, 19, 20],
}

THRESHOLDS = {"timing": 0.30, "motion": 0.25, "handshape": 0.22, "orientation": 0.25}


@dataclass
class FeedbackItem:
    dimension: str        # timing | motion | handshape | orientation
    severity: float       # score in [0,1]
    hand: str             # "left" | "right" | "both" | "-"
    prompt: str           # human-readable corrective instruction


@dataclass
class FeedbackReport:
    target_label: str
    overall_match: float               # 1 - weighted mean score
    items: list = field(default_factory=list)   # FeedbackItem, worst first

    def to_json(self):
        return json.dumps({"target": self.target_label,
                           "overall_match": round(self.overall_match, 3),
                           "items": [asdict(i) for i in self.items]}, indent=1)


def pose_xyz(seq, lm):
    return seq[:, lm * 4: lm * 4 + 3]


def hand_block(seq, base):
    return seq[:, base: base + 63].reshape(len(seq), 21, 3)


def hand_present(seq, base):
    return np.abs(seq[:, base: base + 63]).sum(axis=1) > 0


def dtw_path(a, b):
    """Plain O(n*m) DTW on frame-feature matrices a (n,d), b (m,d)."""
    n, m = len(a), len(b)
    cost = np.linalg.norm(a[:, None, :] - b[None, :, :], axis=2)
    acc = np.full((n + 1, m + 1), np.inf)
    acc[0, 0] = 0
    for i in range(1, n + 1):
        for j in range(1, m + 1):
            acc[i, j] = cost[i - 1, j - 1] + min(acc[i - 1, j], acc[i, j - 1],
                                                 acc[i - 1, j - 1])
    path, (i, j) = [], (n, m)
    while i > 0 and j > 0:
        path.append((i - 1, j - 1))
        k = int(np.argmin([acc[i - 1, j - 1], acc[i - 1, j], acc[i, j - 1]]))
        i, j = (i - 1, j - 1) if k == 0 else (i - 1, j) if k == 1 else (i, j - 1)
    return path[::-1]


def wrist_traj(seq):
    return np.concatenate([pose_xyz(seq, L_WRIST_P), pose_xyz(seq, R_WRIST_P)], axis=1)


def timing_score(path):
    """Sustained tempo deviation from the warping path.

    Tempo is measured only where the GOLD sequence advances — a fast sign
    followed by a static hold otherwise averages back to slope 1 (both
    sequences are fixed 30 frames). ratio < 1 means the learner covered the
    gold's signing phase in fewer frames: too fast.
    Score = |log2(ratio)| clipped to [0,1]; 0.30 ≈ ~25% tempo deviation.
    """
    steps = np.diff(np.array(path), axis=0)
    active = steps[:, 1] > 0
    li, gi = steps[active, 0].sum(), steps[active, 1].sum()
    if li == 0 or gi == 0:
        return 1.0, "faster"
    ratio = li / gi
    return min(abs(float(np.log2(ratio))), 1.0), ("slower" if ratio > 1 else "faster")


def motion_score(learner, gold, path):
    lw, gw = wrist_traj(learner), wrist_traj(gold)
    diffs = np.array([lw[i] - gw[j] for i, j in path])       # (P, 6)
    err = float(np.linalg.norm(diffs, axis=1).mean())
    mean_diff = diffs.mean(axis=0)                            # per-axis bias
    hand = "right" if np.linalg.norm(mean_diff[3:]) >= np.linalg.norm(mean_diff[:3]) else "left"
    d = mean_diff[3:] if hand == "right" else mean_diff[:3]
    dirs = []
    if abs(d[0]) > 0.05: dirs.append("left" if d[0] > 0 else "right")   # learner x too large -> move left (mirrored to signer)
    if abs(d[1]) > 0.05: dirs.append("higher" if d[1] > 0 else "lower")  # image y grows downward
    direction = " and ".join(dirs) if dirs else "closer to the model path"
    return min(err / 0.8, 1.0), hand, direction


def handshape_score(learner, gold, path, base):
    lp, gp = hand_present(learner, base), hand_present(gold, base)
    pairs = [(i, j) for i, j in path if lp[i] and gp[j]]
    if not pairs:
        # gold uses this hand but learner never shows it
        if gp.any() and not lp.any():
            return 1.0, ["hand not detected"]
        return 0.0, []
    lh, gh = hand_block(learner, base), hand_block(gold, base)
    finger_err = {}
    for name, ids in FINGERS.items():
        e = [np.linalg.norm((lh[i, ids] - lh[i, WRIST_H]) - (gh[j, ids] - gh[j, WRIST_H]),
                            axis=1).mean() for i, j in pairs]
        finger_err[name] = float(np.mean(e))
    worst = sorted(finger_err, key=finger_err.get, reverse=True)
    # score on the worst finger, not the mean — a single-finger error
    # (e.g. FOUR vs FIVE differ only in the thumb) must not be diluted
    score = min(finger_err[worst[0]] / 0.30, 1.0)
    fingers = [f for f in worst if finger_err[f] > 0.6 * finger_err[worst[0]]][:2]
    return score, fingers


def orientation_score(learner, gold, path, base):
    lp, gp = hand_present(learner, base), hand_present(gold, base)
    pairs = [(i, j) for i, j in path if lp[i] and gp[j]]
    if not pairs:
        return 0.0
    lh, gh = hand_block(learner, base), hand_block(gold, base)

    def normal(h, t):
        v1 = h[t, INDEX_MCP] - h[t, WRIST_H]
        v2 = h[t, PINKY_MCP] - h[t, WRIST_H]
        n = np.cross(v1, v2)
        return n / (np.linalg.norm(n) + 1e-8)

    angles = [np.degrees(np.arccos(np.clip(np.dot(normal(lh, i), normal(gh, j)), -1, 1)))
              for i, j in pairs]
    return min(float(np.mean(angles)) / 90.0, 1.0)


def compare(learner, gold, target_label):
    """learner, gold: (T, 258) normalized no_face sequences -> FeedbackReport."""
    path = dtw_path(wrist_traj(learner), wrist_traj(gold))
    items = []

    t_score, tempo = timing_score(path)
    if t_score > THRESHOLDS["timing"]:
        items.append(FeedbackItem("timing", round(t_score, 3), "both",
                     f"Your sign is {tempo} than the model — "
                     f"{'take your time' if tempo == 'faster' else 'keep the movement flowing'}."))

    m_score, m_hand, m_dir = motion_score(learner, gold, path)
    if m_score > THRESHOLDS["motion"]:
        items.append(FeedbackItem("motion", round(m_score, 3), m_hand,
                     f"Move your {m_hand} hand {m_dir}."))

    for hand, base in (("left", LHAND0), ("right", RHAND0)):
        h_score, worst = handshape_score(learner, gold, path, base)
        if h_score > THRESHOLDS["handshape"]:
            if worst == ["hand not detected"]:
                items.append(FeedbackItem("handshape", 1.0, hand,
                             f"This sign uses your {hand} hand — keep it visible to the camera."))
            else:
                items.append(FeedbackItem("handshape", round(h_score, 3), hand,
                             f"Check your {hand}-hand shape — adjust your "
                             f"{' and '.join(worst)} finger{'s' if len(worst) > 1 else ''}."))
        o_score = orientation_score(learner, gold, path, base)
        if o_score > THRESHOLDS["orientation"]:
            items.append(FeedbackItem("orientation", round(o_score, 3), hand,
                         f"Rotate your {hand} palm to match the model orientation."))

    items.sort(key=lambda it: it.severity, reverse=True)
    all_scores = [it.severity for it in items] or [0.0]
    return FeedbackReport(target_label, round(1.0 - float(np.mean(all_scores)), 3), items)

File: app/android/app/src/main/kotlin/com/kumpas/kumpas_app/FeedbackEngine.kt
package com.kumpas.kumpas_app

import kotlin.math.abs
import kotlin.math.acos
import kotlin.math.ln
import kotlin.math.min
import kotlin.math.sqrt

/**
 * Kotlin port of training/feedback/feedback_engine.py — DTW-based comparison
 * of a learner sequence against the class gold standard across four
 * dimensions: timing, motion, handshape, orientation.
 *
 * HIGH-RISK CODE (PRD §7, Feedback Algorithm Agent): every behavioural change
 * here must be mirrored in the Python reference and pass the parity unit test
 * (FeedbackEngineParityTest) built from real fixture sequences.
 *
 * Sequences are (T=30) frames of 258 floats (pose 33*4 + left hand 63 +
 * right hand 63), normalized as in build_sequences.py.
 */
object FeedbackEngine {

    const val SEQ_LEN = 30
    const val N_FEATURES = 258
    private const val L_WRIST_P = 15
    private const val R_WRIST_P = 16
    private const val LHAND0 = 132
    private const val RHAND0 = 132 + 63
    private const val WRIST_H = 0
    private const val INDEX_MCP = 5
    private const val PINKY_MCP = 17

    private val FINGERS = linkedMapOf(
        "thumb" to intArrayOf(1, 2, 3, 4),
        "index" to intArrayOf(5, 6, 7, 8),
        "middle" to intArrayOf(9, 10, 11, 12),
        "ring" to intArrayOf(13, 14, 15, 16),
        "pinky" to intArrayOf(17, 18, 19, 20),
    )

    private val THRESHOLDS = mapOf(
        "timing" to 0.30, "motion" to 0.25, "handshape" to 0.22, "orientation" to 0.25
    )

    data class FeedbackItem(
        val dimension: String, val severity: Double, val hand: String, val prompt: String
    )

    data class FeedbackReport(
        val targetLabel: String, val overallMatch: Double, val items: List<FeedbackItem>
    )

    // ---- geometry helpers ----

    private fun poseXyz(seq: Array<FloatArray>, t: Int, lm: Int) =
        doubleArrayOf(seq[t][lm * 4].toDouble(), seq[t][lm * 4 + 1].toDouble(),
                      seq[t][lm * 4 + 2].toDouble())

    private fun handPoint(seq: Array<FloatArray>, t: Int, base: Int, i: Int) =
        doubleArrayOf(seq[t][base + i * 3].toDouble(), seq[t][base + i * 3 + 1].toDouble(),
                      seq[t][base + i * 3 + 2].toDouble())

    private fun handPresent(seq: Array<FloatArray>, t: Int, base: Int): Boolean {
        for (i in 0 until 63) if (seq[t][base + i] != 0f) return true
        return false
    }

    /** 6-dim wrist trajectory frame: left wrist xyz + right wrist xyz. */
    private fun wristFrame(seq: Array<FloatArray>, t: Int): DoubleArray {
        val l = poseXyz(seq, t, L_WRIST_P); val r = poseXyz(seq, t, R_WRIST_P)
        return doubleArrayOf(l[0], l[1], l[2], r[0], r[1], r[2])
    }

    private fun dist(a: DoubleArray, b: DoubleArray): Double {
        var s = 0.0
        for (i in a.indices) { val d = a[i] - b[i]; s += d * d }
        return sqrt(s)
    }

    // ---- DTW (mirrors dtw_path in the Python reference, including tie-break order) ----

    fun dtwPath(a: List<DoubleArray>, b: List<DoubleArray>): List<Pair<Int, Int>> {
        val n = a.size; val m = b.size
        val acc = Array(n + 1) { DoubleArray(m + 1) { Double.POSITIVE_INFINITY } }
        acc[0][0] = 0.0
        for (i in 1..n) for (j in 1..m) {
            acc[i][j] = dist(a[i - 1], b[j - 1]) +
                minOf(acc[i - 1][j], acc[i][j - 1], acc[i - 1][j - 1])
        }
        val path = ArrayList<Pair<Int, Int>>()
        var i = n; var j = m
        while (i > 0 && j > 0) {
            path.add(Pair(i - 1, j - 1))
            // argmin over [diag, up, left] with first-wins ties — same as np.argmin
            val diag = acc[i - 1][j - 1]; val up = acc[i - 1][j]; val left = acc[i][j - 1]
            when {
                diag <= up && diag <= left -> { i--; j-- }
                up <= left -> i--
                else -> j--
            }
        }
        path.reverse()
        return path
    }

    // ---- dimension scores (each mirrors its Python counterpart) ----

    private fun timingScore(path: List<Pair<Int, Int>>): Pair<Double, String> {
        var li = 0.0; var gi = 0.0
        for (k in 1 until path.size) {
            val gStep = path[k].second - path[k - 1].second
            if (gStep > 0) { li += path[k].first - path[k - 1].first; gi += gStep }
        }
        if (li == 0.0 || gi == 0.0) return Pair(1.0, "faster")
        val ratio = li / gi
        val score = min(abs(ln(ratio) / ln(2.0)), 1.0)
        return Pair(score, if (ratio > 1) "slower" else "faster")
    }

    private fun motionScore(
        learner: Array<FloatArray>, gold: Array<FloatArray>, path: List<Pair<Int, Int>>
    ): Triple<Double, String, String> {
        var err = 0.0
        val meanDiff = DoubleArray(6)
        for ((i, j) in path) {
            val lw = wristFrame(learner, i); val gw = wristFrame(gold, j)
            var s = 0.0
            for (k in 0 until 6) { val d = lw[k] - gw[k]; s += d * d; meanDiff[k] += d }
            err += sqrt(s)
        }
        err /= path.size
        for (k in 0 until 6) meanDiff[k] /= path.size
        val normL = sqrt(meanDiff[0] * meanDiff[0] + meanDiff[1] * meanDiff[1] + meanDiff[2] * meanDiff[2])
        val normR = sqrt(meanDiff[3] * meanDiff[3] + meanDiff[4] * meanDiff[4] + meanDiff[5] * meanDiff[5])
        val hand = if (normR >= normL) "right" else "left"
        val d = if (hand == "right") meanDiff.copyOfRange(3, 6) else meanDiff.copyOfRange(0, 3)
        val dirs = ArrayList<String>()
        if (abs(d[0]) > 0.05) dirs.add(if (d[0] > 0) "left" else "right")
        if (abs(d[1]) > 0.05) dirs.add(if (d[1] > 0) "higher" else "lower")
        val direction = if (dirs.isEmpty()) "closer to the model path" else dirs.joinToString(" and ")
        return Triple(min(err / 0.8, 1.0), hand, direction)
    }

    private fun handshapeScore(
        learner: Array<FloatArray>, gold: Array<FloatArray>,
        path: List<Pair<Int, Int>>, base: Int
    ): Pair<Double, List<String>> {
        val pairs = path.filter { (i, j) -> handPresent(learner, i, base) && handPresent(gold, j, base) }
        if (pairs.isEmpty()) {
            val goldUses = (0 until gold.size).any { handPresent(gold, it, base) }
            val learnerUses = (0 until learner.size).any { handPresent(learner, it, base) }
            return if (goldUses && !learnerUses) Pair(1.0, listOf("hand not detected"))
            else Pair(0.0, emptyList())
        }
        val fingerErr = LinkedHashMap<String, Double>()
        for ((name, ids) in FINGERS) {
            var e = 0.0
            for ((i, j) in pairs) {
                val lw = handPoint(learner, i, base, WRIST_H)
                val gw = handPoint(gold, j, base, WRIST_H)
                var fe = 0.0
                for (id in ids) {
                    val lp = handPoint(learner, i, base, id)
                    val gp = handPoint(gold, j, base, id)
                    var s = 0.0
                    for (k in 0 until 3) {
                        val d = (lp[k] - lw[k]) - (gp[k] - gw[k]); s += d * d
                    }
                    fe += sqrt(s)
                }
                e += fe / ids.size
            }
            fingerErr[name] = e / pairs.size
        }
        val worst = fingerErr.entries.sortedByDescending { it.value }
        val score = min(worst[0].value / 0.30, 1.0)
        val fingers = worst.filter { it.value > 0.6 * worst[0].value }.take(2).map { it.key }
        return Pair(score, fingers)
    }

    private fun orientationScore(
        learner: Array<FloatArray>, gold: Array<FloatArray>,
        path: List<Pair<Int, Int>>, base: Int
    ): Double {
        val pairs = path.filter { (i, j) -> handPresent(learner, i, base) && handPresent(gold, j, base) }
        if (pairs.isEmpty()) return 0.0

        fun normal(seq: Array<FloatArray>, t: Int): DoubleArray {
            val w = handPoint(seq, t, base, WRIST_H)
            val im = handPoint(seq, t, base, INDEX_MCP)
            val pm = handPoint(seq, t, base, PINKY_MCP)
            val v1 = doubleArrayOf(im[0] - w[0], im[1] - w[1], im[2] - w[2])
            val v2 = doubleArrayOf(pm[0] - w[0], pm[1] - w[1], pm[2] - w[2])
            val n = doubleArrayOf(
                v1[1] * v2[2] - v1[2] * v2[1],
                v1[2] * v2[0] - v1[0] * v2[2],
                v1[0] * v2[1] - v1[1] * v2[0])
            val len = sqrt(n[0] * n[0] + n[1] * n[1] + n[2] * n[2]) + 1e-8
            return doubleArrayOf(n[0] / len, n[1] / len, n[2] / len)
        }

        var sum = 0.0
        for ((i, j) in pairs) {
            val a = normal(learner, i); val b = normal(gold, j)
            var dot = a[0] * b[0] + a[1] * b[1] + a[2] * b[2]
            dot = dot.coerceIn(-1.0, 1.0)
            sum += Math.toDegrees(acos(dot))
        }
        return min(sum / pairs.size / 90.0, 1.0)
    }

    // ---- public API ----

    fun compare(
        learner: Array<FloatArray>, gold: Array<FloatArray>, targetLabel: String
    ): FeedbackReport {
        val lTraj = (0 until learner.size).map { wristFrame(learner, it) }
        val gTraj = (0 until gold.size).map { wristFrame(gold, it) }
        val path = dtwPath(lTraj, gTraj)
        val items = ArrayList<FeedbackItem>()

        val (tScore, tempo) = timingScore(path)
        if (tScore > THRESHOLDS["timing"]!!) {
            items.add(FeedbackItem("timing", round3(tScore), "both",
                "Your sign is $tempo than the model — " +
                    (if (tempo == "faster") "take your time." else "keep the movement flowing.")))
        }

        val (mScore, mHand, mDir) = motionScore(learner, gold, path)
        if (mScore > THRESHOLDS["motion"]!!) {
            items.add(FeedbackItem("motion", round3(mScore), mHand, "Move your $mHand hand $mDir."))
        }

        for ((hand, base) in listOf("left" to LHAND0, "right" to RHAND0)) {
            val (hScore, worst) = handshapeScore(learner, gold, path, base)
            if (hScore > THRESHOLDS["handshape"]!!) {
                if (worst == listOf("hand not detected")) {
                    items.add(FeedbackItem("handshape", 1.0, hand,
                        "This sign uses your $hand hand — keep it visible to the camera."))
                } else {
                    items.add(FeedbackItem("handshape", round3(hScore), hand,
                        "Check your $hand-hand shape — adjust your ${worst.joinToString(" and ")} " +
                            "finger${if (worst.size > 1) "s" else ""}."))
                }
            }
            val oScore = orientationScore(learner, gold, path, base)
            if (oScore > THRESHOLDS["orientation"]!!) {
                items.add(FeedbackItem("orientation", round3(oScore), hand,
                    "Rotate your $hand palm to match the model orientation."))
            }
        }

        items.sortByDescending { it.severity }
        val mean = if (items.isEmpty()) 0.0 else items.sumOf { it.severity } / items.size
        return FeedbackReport(targetLabel, round3(1.0 - mean), items)
    }

    private fun round3(v: Double) = Math.round(v * 1000.0) / 1000.0
}
